core/src/main/java/org/apache/calcite/sql/fun/SqlTrimFunction.java

```diff
@@ -130,6 +130,9 @@ public SqlTrimFunction(String name, SqlKind kind,
       if (operands[1] == null) {
         operands[1] = SqlLiteral.createCharString(" ", pos);
       }
+      if (operands[2] == null) {
```

Comment:
something is wrong here. If you reached this point you have one argument.

Reply:
I think I've two, right?

For `trim(both 'a' from 'aAa')` you've three `operands`:

```
operand[0]: "BOTH"
operand[1]: "a"
operand[2]: "aAa"
```

Reply:
yes, and the message says that there are "no arguments"

Reply:
Ok so I've debugged through this:

If you call `trim()` the following happens:

- Enters the block for 3 operands (first gets a default value, second and third are `null`, not in the sense of their value, but absent)
- `operands[0]` gets set to `BOTH`
- `operands[1]` gets so to `" "`
- `operands[2]` is `null` (indicating absence of the arg not that it's `null`)

So the actual method call from a user perspective was one without arguments (`trim()`)

`trim(" a ")` works fine
`trim(both " a ")` works fine
`trim("a" from)` syntax error (string to trim is absent)
`trim(both "a" from)` syntax error (string to trim is absent)

So to handle the no-args case I think it's correct to check for `if (operands[2] == null)` in the block with `3` operands. I'll add a comment explaining it, otherwise it's confusing I think.

Reply:
yes, the first two arguments have some default values

Reply:
@mihaibudiu @timgrein Perhaps we can refer to SparkSQL's exception information.
SparkSQL:
```
spark-sql> select trim();
Error in query: Invalid number of arguments for function trim. Expected: one of 1 and 2; Found: 0; line 1 pos 7
```